### ISSUE TITLE
[Feature request] Prometheus : horizontal annotations or instant query for threshold/

### ISSUE THREAD

**Issue by ut0mt8** [grafana/grafana#9085]
My use case is very common : I want to draw a horizontal line on my 95th percentile network graph.

The problem is that prometheus doesn't have function like drawAsInfinite or something like this.

Digging into options I see too possibilities : 

- possibilites to make "horizontal annotations" based on a promql query returning a single value (instant query) 
- and/or add the possibility to use an instant query for a threshold.

**Comment by torkelo** [comment]
changed name as this https://github.com/grafana/grafana/issues/9084 covers the other feature request

**Comment by davkal** [comment]
@ut0mt8 let me understand your use case a bit better. What's the benefit of drawing a straight line representing the latest 95th percentile, overlaying it to the whole range with contains data points at which the value of that 95th percentile was very likely different.

**Comment by RichiH** [comment]
The usual use case in network is 95/5 accounting: https://en.wikipedia.org/wiki/Burstable_billing

You take 5 minute buckets (starting at 00:00 and following), throw away what's above the 95th percentile, and bill what's left. Being able to see this is important when checking bill disputes.

**Comment by ut0mt8** [comment]
Yes absolutely. This is definitively a must have for grafana if we want to use it for billing network graph.
This is not only drawing a straight line , but also calculate it. But the compute of this value should be done with the tsdb, and I don't find a way to do it (either prometheus or influx). This is deceptive since RRDtool do this from the beginning. (and even mrtg)

**Comment by simPod** [comment]
I have achieved it in postgres datasource by generating time series data and joining our network  commitments on it in a metric. Not sure about influx tho.

**Comment by elacour** [comment]
I'm interested too, I draw river water height and I want to put static horizontal lines for well known river floods. (I use influxdb)

**Comment by swaranga** [comment]
+1 

For an availability graph, I would like to draw horizontal lines reminding different severities of the issue if availability is below a certain threshold.

I feel this should be a common feature for all Graph panels irrespective of the datasource.

**Comment by danielfaust** [comment]
+1

I want to draw a simple horizontal line at 50 Hz in a power line frequency graph, as well as a 230 V in a power line voltage graph, in order to have a proper reference line.

**Comment by torkelo** [comment]
Use graph threshold for that

**Comment by danielfaust** [comment]
Thanks @torkelo 

Are there any plans to implement a proper solution or is this workaround the way to go? It works, which is great, but it doesn't feel right.

**Comment by torkelo** [comment]
What are you missing?

**Comment by ut0mt8** [comment]
I will check, but I don't think this will fit for 95th line. What I need is an horizontal line with a value calculated over the time range selected. So the horizontal line value should be calculated from a special metric / expression. That said I don't think it's even possible to calculate this value from promql (which my backend of choice). For example of graph see https://en.wikipedia.org/wiki/Burstable_billing for example.

**Comment by nanicpc** [comment]
Current thresholds are static, and in many cases, as the one of 95th percentile, are simply not enough.
I give you the example of my company. We have different thresholds according to a set of conditions (parameters / templating variables). thresholds are stored in some db (time invariable) in a table like this:

Use_Case|Parameter_level1|Parameter_level2|Threshold
------------ | ------------- | ------------- | -------------
server_limit|ip|shift1|50
server_limit|ip|shift2|30
working_hours|country|work_day|7

Having thresholds or horizontal annotations with queries will resolve this issue, because that could allow us to automatically update the threshold according to the use case chosen by the user.

**Comment by mcclauba** [comment]
@torkelo You mentioned that thresholds work as a straight line, which is true.  However I require alarms on my graphs as well, which disables thresholds.  My work around is writing a constant value to my database, but this seems very wasteful.

**Comment by coderanger** [comment]
I think https://github.com/grafana/grafana/pull/19102 also addresses at least some of this? You can use an Instant query where the expression is a static value or a 95th percentile or whatever and then use the `constant` transform to turn that into a line on the graph.

**Comment by anarcat** [comment]
> I think #19102 also addresses at least some of this? You can use an Instant query where the expression is a static value or a 95th percentile or whatever and then use the constant transform to turn that into a line on the graph.

nice! that does looks like it fixes this problem, at least for my use case (95% percentiles).

**Comment by bob454522** [comment]
hi,  i see that the github PR implements this in grafana when using Prometheus,  however is there anyway to make use of this when using Graphite/WhisperDB as the Query Source?

(not including installing / adding a Prometheus instance for this one usecase).

i would really love to have the ability to make a horizontal line on some graphs, and while using the grafana (w graphite), and while the Threshold setting does do what i need,  Threshold adds the slider which kills the horizantal width of your graph  (especially when adding multiple horzontal lines via thresholds)
thanks

**Comment by MikeC-BC** [comment]
Please let us just have a horizontal line if there is a single data point without any hacks such as graph threshold or extrapolating a single data point to form a time series.

**Comment by peterdk** [comment]
I have some daily data from Influxd, and Grafana already shows the mean/min/max in the legend table. However, there is no way to use that calculation in the chart with a horizontal line. I want to display the mean line based on the data in the chart, and weirdly I just cannot do that. I would think since the calculations are already there, it should be easy to add?

**Comment by AlexOConnorHub** [comment]
While not dynamic, as this feature request is asking for, (and most defiantly not idea), a user on the [grafana form](https://community.grafana.com/t/draw-static-lines/8076/4) shared that you can set a threshold and display the line. It may work for people in a pinch, like my use case of drawing a static line showing the number of CPUs.

**Comment by AlexOConnorHub** [comment]
> I have some daily data from Influxd, and Grafana already shows the mean/min/max in the legend table. However, there is no way to use that calculation in the chart with a horizontal line. I want to display the mean line based on the data in the chart, and weirdly I just cannot do that. I would think since the calculations are already there, it should be easy to add?

In influx 2.6 (the flux query language), I just found a solution for graphing a constant. The important part of the query is the `aggregateWindow` line. You can pass any function, including built-in or user-defined, while allowing for the timestamp to be maintained, allowing it to be graphed across the graph. In my case, I have a [` node_exporter`](https://github.com/prometheus/node_exporter) instance I am scraping into Influx.

```
from(bucket: v.bucket)
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r["_measurement"] == "node_cpu_seconds_total")
  |> filter(fn: (r) => r["_field"] == "counter")
  |> group(columns: ["cpu"])
  |> aggregateWindow(every: v.windowPeriod, fn: count, createEmpty: false)
  |> yield()
```

It would be nice if, in a grafana override, we could select the output of the query (say, just the `count()` result), and specify that it should be graphed over the entire time period.

**Comment by WhoisUnknown** [comment]
Hi, Any updates on this feature?

**Comment by guimazoni** [comment]
Status?
